File: Java/Linked List Cycle II.java
// https://leetcode.com/problems/linked-list-cycle-ii/

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
        ListNode fast=head;
        ListNode slow=head;
        boolean flag=false;
        while(fast!=null && fast.next!=null && slow!=null){
            fast=fast.next.next;
            slow=slow.next;
            if(slow==fast){
                flag=true;
                break;
            }
        }
        if(flag){
            slow=head;
            while(slow!=fast){
                slow=slow.next;
                fast=fast.next;
            }
            return slow;
        }else{
            return null;
        }
    }
}

File: Python/Linked_List_Cycle_II.py
#LeetCode problem link: https://leetcode.com/problems/linked-list-cycle-ii/


class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        node = head
        val_holder = []
        while(node.next is not None):
            temp = node.val
            val_holder.append(temp)
            if  node.next.val not in val_holder:
                node = node.next
            else:
                return node.next
